From 713128d438de8a1f69be70d89666cf3f02678aa2 Mon Sep 17 00:00:00 2001
From: Alexey Makhalov <alexey.makhalov@broadcom.com>
Date: Thu, 24 Oct 2024 08:48:23 +0000
Subject: [PATCH] NEWCA: make initrd pages immutable

NEWCA initrd images can be shared across CRX instances. And there are no
cases where initrd files can be modified. From other side, accidental
modifications can lead to expensive COWs from the hypervisor side.
Seal these files to make corresponding NEWCA initrd pages immutable.

Signed-off-by: Alexey Makhalov <alexey.makhalov@broadcom.com>
---
 init/initramfs.c | 6 ++++++
 mm/shmem.c       | 4 +++-
 2 files changed, 9 insertions(+), 1 deletion(-)

diff --git a/init/initramfs.c b/init/initramfs.c
index 06e251747..ded0ea74f 100644
--- a/init/initramfs.c
+++ b/init/initramfs.c
@@ -12,6 +12,7 @@
 #include <linux/utime.h>
 #include <linux/file.h>
 #include <linux/memblock.h>
+#include <linux/memfd.h>
 #include <linux/mm.h>
 #include <linux/namei.h>
 #include <linux/init_syscalls.h>
@@ -505,6 +506,11 @@ static int __init do_name(void)
 			vfs_fchmod(wfile, mode);
 			if (body_len)
 				vfs_truncate(&wfile->f_path, body_len);
+			/* Disallow modifications of NEWCA pages. */
+			if (page_aligned_fmt)
+				BUG_ON(memfd_fcntl(wfile, F_ADD_SEALS,
+					F_SEAL_SEAL | F_SEAL_SHRINK |
+					F_SEAL_GROW | F_SEAL_WRITE));
 			state = CopyFile;
 		}
 	} else if (S_ISDIR(mode)) {
diff --git a/mm/shmem.c b/mm/shmem.c
index f7c08e169..5d68abe55 100644
--- a/mm/shmem.c
+++ b/mm/shmem.c
@@ -2365,7 +2365,9 @@ static struct inode *shmem_get_inode(struct super_block *sb, struct inode *dir,
 		memset(info, 0, (char *)inode - (char *)info);
 		spin_lock_init(&info->lock);
 		atomic_set(&info->stop_eviction, 0);
-		info->seals = F_SEAL_SEAL;
+		/* Allow sealing while extracting initrd. */
+		if (system_state >= SYSTEM_FREEING_INITMEM)
+			info->seals = F_SEAL_SEAL;
 		info->flags = flags & VM_NORESERVE;
 		info->i_crtime = inode->i_mtime;
 		info->fsflags = (dir == NULL) ? 0 :
-- 
2.39.4

